### PR TITLE
Added test for F-9H

### PR DESCRIPTION
**Pull request by pragamj**
Signed-off-by: pragam <pragam.jain@seagate.com>

# Problem Statement
- Added test: [TEST-41953][TEST-42176]

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [x] Attach test execution logs
[TEST-41953_TestK8CortxUpgrade_20220608101330.log](https://github.com/Seagate/cortx-test/files/8871018/TEST-41953_TestK8CortxUpgrade_20220608101330.log)
[TEST-42176_TestK8CortxUpgrade_20220608080400.log](https://github.com/Seagate/cortx-test/files/8871019/TEST-42176_TestK8CortxUpgrade_202206080804
[TEST-42179_TestK8CortxUpgrade_20220610035219.log](https://github.com/Seagate/cortx-test/files/8888231/TEST-42179_TestK8CortxUpgrade_20220610035219.log)
00.log)

-  [x] Collection tested and no collection error introduced (`pytest --local True --collect-only`)
[collection_log.log](https://github.com/Seagate/cortx-test/files/8871020/collection_log.log)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: comptests/prov/test_prov_k8s_cortx_cold_upgrade.py [hunk 1]
```diff
@@ -184,10 +185,10 @@ def test_37352(self):
                     "Step 6: Start upgrade and verify pod went into crashloopbackoffstate or not")
                 LOGGER.info("Upgrading CORTX image to version: %s.",
                             self.cortx_all_image)
-                resp = self.deploy_lc_obj.upgrade_software(self.master_node_obj,
-                                                           self.prov_deploy_cfg["git_remote_path"],
-                                                           upgrade_type="cold")
+                resp = self.upgrade_lc_obj.upgrade_software(self.master_node_obj,
```
Comment (pranav-sahasrabudhe) [inline comptests/prov/test_prov_k8s_cortx_cold_upgrade.py:188]: We should preferably avoid `lc` word in the variable names. If its a lot of refactoring and needed across all the code then we might want to differ it. But if its in a small context, let's do it.

**Review by pranav-sahasrabudhe — CHANGES_REQUESTED** [review]
The changes requested are optional. Please decide and do the needful.

File: comptests/prov/test_prov_k8s_cortx_cold_upgrade.py (hunk 1 above)
Comment (pragamj) [reply comptests/prov/test_prov_k8s_cortx_cold_upgrade.py]: done

File: comptests/prov/test_prov_k8s_cortx_cold_upgrade.py [hunk 2]
```diff
@@ -75,58 +74,58 @@ def test_37357(self):
...
-        installed_version = self.deploy_lc_obj.get_installed_version(self.master_node_obj)
+        LOGGER.info("Step 1: Get installed version.")
+        installed_version = self.deploy_obj.get_installed_version(self.master_node_obj)
         LOGGER.info("Current version: %s", installed_version)
         LOGGER.info("Step 1: Done.")
-        LOGGER.info(
-            "Step 2: Check if installing version is higher than installed version.")
-        self.deploy_lc_obj.generate_and_compare_version(self.cortx_all_image, installed_version)
+        LOGGER.info("Step 2: Check if installing version is higher than installed version.")
+        self.deploy_obj.generate_and_compare_both_version(self.cortx_all_image,
```
Comment (priyankaborawake) [inline comptests/prov/test_prov_k8s_cortx_cold_upgrade.py:82]: We should asserting if the function returns false.

**Comment by Kulkarni-Kaustubh** [comment]
Verified

**Review by pranav-sahasrabudhe — APPROVED** [review]
Approved

**Review by sonalk0209 — APPROVED** [review]
Approving for merge!!!

**Review by ravindrachoudhari — APPROVED** [review]

File: comptests/prov/test_prov_k8s_cortx_cold_upgrade.py [hunk 3]
```diff
@@ -75,58 +74,59 @@ def test_37357(self):
...
+                                                 upgrade_type="cold")
         assert_utils.assert_true(resp[0], resp[1])
         LOGGER.info("Step 6: Done.")
 
-        LOGGER.info(
-            "Step 7: Check if installed version is equals to installing version.")
-        installed_version = self.deploy_lc_obj.generate_and_compare_version(self.cortx_all_image, installed_version)
+        LOGGER.info("Step 7: Check if installed version is equals to installing version")
+        installed_version = self.deploy_obj.generate_and_compare_both_version(
+            self.cortx_all_image, installed_version)
```
Comment (priyankaborawake) [inline comptests/prov/test_prov_k8s_cortx_cold_upgrade.py:129]: Does this method check for equal version ? I guess not.

File: comptests/prov/test_prov_k8s_cortx_granular.py [hunk 4]
```diff
@@ -346,7 +353,7 @@ def test_37356(self):
         version = resp[1].split("cortx-rgw:")[1].split("-")
         new_version = version[0] + "-" + version[1].strip()
         LOGGER.info(new_version)
-        installing_version = self.deploy_lc_obj.generate_and_compare_both_version(
+        installing_version = self.deploy_obj.generate_and_compare_both_version(
```
Comment (priyankaborawake) [inline comptests/prov/test_prov_k8s_cortx_granular.py:356]: Add assert here.

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 5]
```diff
@@ -78,7 +82,22 @@ def teardown_class(self):
             path = os.path.join(LOG_DIR, LATEST_LOG_FOLDER)
             support_bundle_utils.collect_support_bundle_k8s(local_dir_path=path,
                                                             scripts_path=
-                                                            self.deploy_conf['k8s_dir'])
+                                                            self.prov_conf['k8s_dir'])
+
+    def rolling_upgrade(self, exc: bool = True, output=None, flag=None):
+        """Function calls upgrade and put return value in queue object."""
+        resp = self.upgrade_obj.upgrade_software(self.master_node_list[0],
```
Comment (priyankaborawake) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:89]: add params in docstring.

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 6]
```diff
@@ -154,9 +175,201 @@ def test_33669(self):
...
+        # Pull upgrade Images on all worker nodes
+        for each in self.worker_node_list:
+            worker_thread = Thread(target=self.upgrade_obj.prov_obj.pull_cortx_image,
+                                   args=(each,))
+            worker_thread.start()
+            pull_image_thread_list.append(worker_thread)
+        for each in pull_image_thread_list:
+            each.join()
+        que = multiprocessing.Queue()
+        upgrade_thread = multiprocessing.Process(target=self.rolling_upgrade, args=(
```
Comment (priyankaborawake) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:268]: this should be upgrade_process,

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 7]
```diff
@@ -154,9 +175,201 @@ def test_33669(self):
...
+            True, que, self.prov_conf["upgrade_start"]))
+        suspend_upgrade_thread = multiprocessing.Process(target=self.rolling_upgrade, args=(
```
Comment (priyankaborawake) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:270]: suspend_upgrade_process

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 8]
```diff
@@ -154,9 +175,201 @@ def test_33669(self):
...
+        pull_image_thread_list = []
+        for each in self.worker_node_list:
+            worker_thread = Thread(target=self.upgrade_obj.prov_obj.pull_cortx_image, args=(each
+                                                                                               ,))
+            worker_thread.start()
+            pull_image_thread_list.append(worker_thread)
+        for each in pull_image_thread_list:
+            each.join()
+        que = multiprocessing.Queue()
+        upgrade_thread = multiprocessing.Process(target=self.rolling_upgrade, args=(
```
Comment (priyankaborawake) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:334]: same as above.

File: comptests/prov/test_prov_k8s_cortx_cold_upgrade.py (hunk 3 above)
Comment (pragamj) [reply comptests/prov/test_prov_k8s_cortx_cold_upgrade.py:129]: No

**Review by priyankaborawake — APPROVED** [review]

**Pull request merged**